### DISCUSSION TITLE
Expose port from TestContainer built from Dockerfile

### DISCUSSION THREAD

**Discussion by george-petrakis** [testcontainers/testcontainers-dotnet#1001] [Q&A]
Hey, and thanks for all the hard work! How can I expose lets say port 1433 on the following:

```     
   private readonly IFutureDockerImage database = 
            new ImageFromDockerfileBuilder()
                .WithDockerfileDirectory(Path.Combine(CommonDirectoryPath.GetSolutionDirectory().DirectoryPath, "../"))
                .WithDockerfile("docker/Dockerfile")
                .Build();
```

I've tried WithPortForwarding but to no avail, same goes with passing Run arguments , I can't find a way of doing that, cheers!

**Comment by HofmeisterAn** [comment]
I am a bit confused. You need to publish (bind) ports using the `ContainerBuilder`, not the `ImageFromDockerfileBuilder`. Are you referring to [EXPOSE](https://docs.docker.com/engine/reference/builder/#expose)?

> The `EXPOSE` instruction does not actually publish the port. It functions as a type of documentation between the person who builds the image and the person who runs the container, about which ports are intended to be published.  

Publishing (binding) the port works regardless if it is "exposed" via `EXPOSE` from the Dockerfile or not, you can simply use `_ = new ContainerBuilder().WithPortBinding(1433, true)`. If you still want to set `EXPOSE` in the Dockerfile, you can use `WithBuildArgument(string, string)` and an `ARG` (variable).

**Reply by george-petrakis** [reply]
Let me get this straight , I've built my image using ImageFromDockerfileBuilder, how will I proceed to run said image with a host port exposed?

**Reply by HofmeisterAn** [reply]
No additional configuration is required; binding a port works in the same manner. You will find more information about network communication in our [documentation](https://dotnet.testcontainers.org/api/create_docker_network/).

```csharp
var futureImage = new ImageFromDockerfileBuilder()
    .WithDockerfileDirectory(CommonDirectoryPath.GetSolutionDirectory(), string.Empty)
    .WithDockerfile("Dockerfile")
    .Build();

await futureImage.CreateAsync()
    .ConfigureAwait(false);

var container = new ContainerBuilder()
    .WithImage(futureImage)
    .WithPortBinding(containerPort, true)
    .Build();

await container.StartAsync()
    .ConfigureAwait(false);

_ = container.GetMappedPublicPort(containerPort);
```
